### PR TITLE
docs: [IAC-2021]: Document Terraform backend initialization

### PR DESCRIPTION
**Pull request by richardblack-Harness**
## Description

* Document the use of environment variables to initialize the terraform backend to populate the `terraform init -backend-config` parameters at runtime
* [JIRA](https://harness.atlassian.net/browse/IAC-2021)

### Screenshot / Preview
RemoteBackends

PRs must meet these requirements to be merged:

- [x] Successful preview build.
- [x] Code owner review.
- [x] No merge conflicts.
- [x] Release notes/new features docs: Feature/version released to at least one prod environment.

### CODE REVIEW

File: docs/infra-as-code-management/pipelines/iacm-plugins/terraform-plugins.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ The supported commands include:
 
 ### Initialize
 The `terraform init` command prepares a working directory containing Terraform configuration files, performing key functions like:
-- **Backend Initialization**: Sets up the backend for Terraform state management.
+- **Backend Initialization**: Sets up the backend for Terraform state management. Go to [Terraform backend initialization](https://developer.harness.io/docs/infra-as-code-management/terraform-backend/init-configuration) for more information on how to use environment variables to dynamically initialize your infrastructure.
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/pipelines/iacm-plugins/terraform-plugins.md:30]:
```suggestion
- **Backend Initialization**: Sets up the backend for Terraform state management. Go to [Terraform backend initialization](https://developer.harness.io/docs/infra-as-code-management/terraform-backend/init-configuration) for more information on how to use environment variables to initialize your infrastructure dynamically.
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+---
+title: Initialization
+description: Initialize your infrastructure dynamically with environment variables
+sidebar_position: 10
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Dynamic Backend Configuration
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:10]:
```suggestion
# Dynamic backend configuration
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  backend "s3" {
+    bucket = "mybucket"
+    key    = "path/to/my/key"
+    region = "us-east-1"
+  }
+}
+```
+
+:::tip
+While this approach works, it lacks flexibility and can be challenging to update or manage in different environments (e.g., development, staging, production).
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:30]:
```suggestion
While this approach works, it lacks flexibility and can be challenging to update or manage in different environments (for example, development, staging, production).
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+:::
+
+## Apply environment variables
+Using environment variables allows for a more dynamic and flexible backend configuration. By specifying an empty or partially configured backend block, we can populate the necessary values through environment variables.
+
+### Partial Configuration Example
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:36]:
```suggestion
### Partial configuration example
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+```hcl
+terraform {
+  backend "s3" {}
+}
+```
+
+In this example, the backend configuration is left empty, allowing us to supply the values using environment variables during the `terraform init` command.
+
+## Setting Environment Variables
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:45]:
```suggestion
## Set environment variables
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+The environment variables used to configure the backend can be set in various ways, such as directly in the terminal or through CI/CD pipelines. Here are the variables corresponding to the partial configuration:
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:46]:
```suggestion
You can set the environment variables used to configure the backend in various ways, such as directly in the terminal or through CI/CD pipelines. Here are the variables corresponding to the partial configuration:
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+To set up an environment variable that population the terraform backend configuration as part of the `terraform init -backend-config` command, for example:
+
+`terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`
+
+You must name your variables in the following format:
+
+`PLUGIN_PLUGIN_INIT_BACKEND_CONFIG_` followed by a unique identifier to distinguish your variable.
+:::
+
+When initializing an AWS S3 bucket for example, and passing the `bucket`, `key` and `region` parameters, Harness recommends naming your variables similar to the following:
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:58]:
```suggestion
When initializing an AWS S3 bucket for example, and passing the `bucket`, `key`, and `region` parameters, Harness recommends naming your variables similar to the following:
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+}
+```
+
+In this example, the backend configuration is left empty, allowing us to supply the values using environment variables during the `terraform init` command.
+
+## Setting Environment Variables
+The environment variables used to configure the backend can be set in various ways, such as directly in the terminal or through CI/CD pipelines. Here are the variables corresponding to the partial configuration:
+
+:::important
+To set up an environment variable that population the terraform backend configuration as part of the `terraform init -backend-config` command, for example:
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:49]:
```suggestion
To set up an environment variable that populates the terraform backend configuration as part of the `terraform init -backend-config` command (for example: `terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`), you must name your variables in the following format:
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+`terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:51]:
```suggestion
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+You must name your variables in the following format:
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:53]:
```suggestion
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+`PLUGIN_PLUGIN_INIT_BACKEND_CONFIG_` followed by a unique identifier to distinguish your variable.
+:::
+
+When initializing an AWS S3 bucket for example, and passing the `bucket`, `key` and `region` parameters, Harness recommends naming your variables similar to the following:
+- `PLUGIN_INIT_BACKEND_CONFIG_BUCKET`
+- `PLUGIN_INIT_BACKEND_CONFIG_KEY`
+- `PLUGIN_INIT_BACKEND_CONFIG_REGION`
+
+
+## Adding New Environment Variables
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:64]:
```suggestion
## Add new environment variables
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+Once your terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.
+
+![Pipeline environment variable usage](./static/tf-initi-variable-log.png)
+
+Notice the list of backend config values, shortly followed by the generated `terraform init` command including the `-backend-config` label with accompanying parameters, populated with your environment variable values.
+
+
+## Conclusion
+Using environment variables for backend configuration in Terraform provides a flexible and dynamic approach, making it easier to manage different environments and improve security. By following the steps outlined in this document, you can set up and use environment variables to configure your Terraform backend dynamically.
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:105]:
```suggestion
Using environment variables for backend configuration in Terraform provides a flexible and dynamic approach, making it easier to manage different environments and improve security. By following the steps outlined in this topic, you can set up and use environment variables to configure your Terraform backend dynamically.
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+  ![Terraform init environment variables](./static/tf-init-env-variables.png)
+  </TabItem>
+</Tabs>
+
+
+
+## Review variable usage
+
+Once your terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:97]:
```suggestion
Once your Terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md (hunk 4 above)
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:36]: Recommend using sentence case for all headings.

https://harness.atlassian.net/wiki/spaces/DOCS/pages/21544468526/Harness+documentation+style+guide#Headings%2C-titles

File: docs/infra-as-code-management/terraform-backend/init-configuration.md (hunk 5 above)
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:45]: Recommend avoiding gerunds for all headings.

https://harness.atlassian.net/wiki/spaces/DOCS/pages/21544468526/Harness+documentation+style+guide#Headings%2C-titles

**Review by brian-f-harness — APPROVED** [review]
Looks good. Added a few suggestions, mostly around the headings style guide.

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+In this example, the backend configuration is left empty, allowing us to supply the values using environment variables during the `terraform init` command.
+
+## Set environment variables
+You can set the environment variables used to configure the backend in various ways, such as directly in the terminal or through Infrastructure pipelines. Here are the variables corresponding to the partial configuration:
+
+:::important
+To set up an environment variable that populates the terraform backend configuration as part of the `terraform init -backend-config` command (for example: `terraform init -backend-config="bucket=sample_s3_bucket" -backend-config="key=terraform.tfstate" -backend-config="region=us-east-1"`), you must name your variables in the following format:
+
+
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:51]:
```suggestion
```

File: docs/infra-as-code-management/terraform-backend/init-configuration.md [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+## Review variable usage
+
+Once your Terraform backend configuration environment variables are set and you have a provision pipeline, run the pipeline and review the console log during the `init` step to confirm your variable values match what is used in the pipeline.  
+
+![Pipeline environment variable usage](./static/tf-initi-variable-log.png)
+
+Notice the list of backend config values, shortly followed by the generated `terraform init` command including the `-backend-config` label with accompanying parameters, populated with your environment variable values.
+
```
Comment (brian-f-harness) [inline docs/infra-as-code-management/terraform-backend/init-configuration.md:100]:
```suggestion
```

**Review by brian-f-harness — DISMISSED** [review]
LGTM

File: docs/infra-as-code-management/terraform-backend/_category_.json [hunk 16]
```diff
@@ -0,0 +1,14 @@
+{
+    "label":"Terraform Backend",
```
Comment (dwarren-dev) [inline docs/infra-as-code-management/terraform-backend/_category_.json:2]: although this is technically a remote backend, opentofu suppports this too so lets call this "Remote Backends"

Comment (richardblack-Harness) [reply docs/infra-as-code-management/terraform-backend/_category_.json:2]: Fair enough

**Review by brian-f-harness — DISMISSED** [review]

**Review by dwarren-dev — APPROVED** [review]

**Pull request merged**